[
    {
        "title": "Too many vdev probe errors should suspend pool"
    },
    {
        "author": {
            "login": "PR Description"
        },
        "body": "### Motivation and Context\r\nSimilar to what we saw in #16569, we need to consider that a replacing vdev should not be considered as fully contributing to the redundancy of a raidz vdev even though current IO has enough redundancy.  I have seen raidz3 pools where there were 4 missing disks (one involved in a replacing vdev) and the pool was still online and taking IO.  This case is different from #16569 in that ZED was not running so the vdev_probe() errors are driving the diagnosis here.\r\n\r\n### Description\r\nWhen a failed vdev_probe() is faulting a disk, it now checks if that disk is required, and if so it suspends the pool until the admin can return the missing disks. \r\n\r\nSponsored-by: Klara, Inc. \r\nSponsored-by: Wasabi Technology, Inc. \r\n\r\n ### How Has This Been Tested?\r\nAdded a new test that verifies that probe errors from 4 disks on a raidz3 with a replacing vdev will suspend the pool.  Before the change the pool would not suspend.\r\n```\r\n  pool: testpool \r\nstate: SUSPENDED \r\nstatus: One or more devices are faulted in response to IO failures. \r\naction: Make sure the affected devices are connected, then run 'zpool clear'. \r\n   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-HC \r\n  scan: resilvered 6.47M in 00:00:03 with 8198 errors on Fri Dec 13 16:55:12 2024 \r\nconfig: \r\n        NAME                  STATE     READ WRITE CKSUM \r\n        testpool              ONLINE       0     0     0 \r\n          raidz3-0            ONLINE       0     0     0 \r\n            /var/tmp/dev-0    ONLINE       0     0 8.16K \r\n            /var/tmp/dev-1    ONLINE       0     0    70 \r\n            /var/tmp/dev-2    ONLINE       0     0    61 \r\n            /var/tmp/dev-3    ONLINE       0     0    48 \r\n            /var/tmp/dev-4    ONLINE       0     0    55 \r\n            /var/tmp/dev-5    ONLINE       0     0    62 \r\n            /var/tmp/dev-6    ONLINE       0     0 8.11K \r\n            sdh1              DEGRADED   195   707     0  too many errors \r\n            sdh2              DEGRADED   278 4.95K     0  too many errors \r\n            sdh3              DEGRADED   363 9.05K     0  too many errors \r\n            replacing-10      ONLINE       0     0 28.3K \r\n              sdh4            DEGRADED   453 41.5K     0  too many errors \r\n              /var/tmp/dev-7  ONLINE       0     0     0 \r\n \r\nerrors: 8168 data errors, use '-v' for a list \r\n```\r\n\r\n### Types of changes\r\n- [x] Bug fix (non-breaking change which fixes an issue)\r\n- [ ] New feature (non-breaking change which adds functionality)\r\n- [ ] Performance enhancement (non-breaking change which improves efficiency)\r\n- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)\r\n- [ ] Breaking change (fix or feature that would cause existing functionality to change)\r\n- [ ] Library ABI change (libzfs, libzfs\\_core, libnvpair, libuutil and libzfsbootenv)\r\n- [ ] Documentation (a change to man pages or other documentation)\r\n\r\n### Checklist:\r\n- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).\r\n- [ ] I have updated the documentation accordingly.\r\n- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).\r\n- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.\r\n- [ ] I have run the ZFS Test Suite with this change applied.\r\n- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by)."
    },
    {
        "author": {
            "login": "allanjude"
        },
        "body": "Reviewed-by: Allan Jude <<allan@klarasystems.com>>"
    },
    {
        "author": {
            "login": "behlendorf"
        },
        "body": ""
    },
    {
        "author": {
            "login": "amotin"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "tonyhutter"
        },
        "body": ""
    },
    {
        "author": {
            "login": "amotin"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "amotin"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "amotin"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "behlendorf"
        },
        "body": ""
    },
    {
        "author": {
            "login": "behlendorf"
        },
        "body": ""
    },
    {
        "author": {
            "login": "tonyhutter"
        },
        "body": ""
    },
    {
        "author": {
            "login": "behlendorf"
        },
        "body": ""
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": ""
    },
    {
        "author": {
            "login": "amotin"
        },
        "body": "@don-brady CI is very unhappy on `fault/suspend_on_probe_errors` test."
    },
    {
        "author": {
            "login": "behlendorf"
        },
        "body": ">  CI is very unhappy on fault/suspend_on_probe_errors test.\r\n\r\nIt seems to be detecting pool errors in this failure case.  That's concerning but I wonder if it's somehow mistaken.\r\n\r\n```\r\n  21:07:33.66 SUCCESS: zpool status -v testpool\r\n  21:07:33.70 ERROR: check_pool_status testpool errors No known data errors exited 1\r\n  ```"
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": "So some of the test failures were from the fact that  the resilver had not completed when the scrub was requested so the scrub request failed.  That was fixed by zpool wait -t resilver.  The other case is that after the scrub, some files were still showing as being corrupted.  However, if the pool is scrubbed a second time, those errors go away.\r\n\r\nI've observed this phenomena in practice but not sure the underlying reason."
    },
    {
        "author": {
            "login": "don-brady"
        },
        "body": "Addressed recent feedback and rebased to latest master"
    },
    {
        "data": {
            "repository": {
                "issue": null
            }
        },
        "errors": [
            {
                "type": "NOT_FOUND",
                "path": [
                    "repository",
                    "issue"
                ],
                "locations": [
                    {
                        "line": 4,
                        "column": 9
                    }
                ],
                "message": "Could not resolve to an Issue with the number of 16569."
            }
        ]
    }
]